### ISSUE TITLE
Feature Request: Support IRSA in AWS auth method

### ISSUE THREAD

**Issue by lonex-branch** [hashicorp/terraform-provider-vault#1167]
The latest version of this provider, v2.23.0, doesn't support IAM role (for Kubernetes service account) authentication in AWS auth method. However, it works in the latest version of the Vault cli, v1.8.2.

Context

We want to use IRSA to authenticate Vault for Kubernetes service. The Vault server is external -- not deployed in kubernetes. The IRSA has been working for the terraform provider aws in sts assume'ing IAM role. We want to use the aws web identity token to auth against vault. The k8s container has the IRSA env settings:

```
AWS_DEFAULT_REGION:           us-west-1
AWS_REGION:                   us-west-1
AWS_ROLE_ARN:                 arn:aws:iam::<aws-account-id>:role/dev_role
AWS_WEB_IDENTITY_TOKEN_FILE:  /var/run/secrets/eks.amazonaws.com/serviceaccount/token
```

We've tried different variation of the vault provider `auth_login` configure block and nothing worked, e.g. 

```
provider "vault" {
  address = var.vault_addr
  auth_login {
    path = "auth/aws/login"
    parameters = {
      role                 = var.vault_role
    }
  }
}
```

The vault cli works using the web identity token as it can retrieve the vault token after the successful auth

```
vault login -method=aws role=dev_role
```

Some details

We've noticed that the provider depends on v1.2.0 of vault https://github.com/hashicorp/terraform-provider-vault/blob/v2.23.0/go.mod#L21. The so-called WebIdentityRoleProvider is not part of the provider chain config https://github.com/hashicorp/vault/blob/v1.2.0/helper/awsutil/generate_credentials.go#L37, while the master version of the vault has the WebIdentityRoleProvider https://github.com/hashicorp/vault/blob/master/sdk/helper/awsutil/generate_credentials.go#L103.

The natural idea is to upgrade the vault lib dependency. Is that possible for the Vault provider to upgrade its vault dependency to the latest. The WebIdentityRoleProvider has been implemented in vault lib for about a year already.

### References

IRSA: https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html

**Comment by bakeemawaytoys** [comment]
I did some digging to see I could add this feature, but I think it has already been added.  Support for the web identity provider was added in Vault 1.4.3 with PR [7738](https://github.com/hashicorp/vault/pull/7738).  In PR [12128](https://github.com/hashicorp/vault/pull/12128) Vault's AWS authentication code was replaced with the [hashicorp/go-secure-stdlib](https://github.com/hashicorp/go-secure-stdlib) library.  It looks like the provider was updated to use the go-secure-stdlib in #1259 and the change was released in version 3.1.0.  One thing [I found while digging through the code](https://github.com/hashicorp/terraform-provider-vault/blob/bd9e45f53d03cd257a0fb01592b17718b969c069/vault/provider.go#L1026) is that it is possible to configure the region used to sign the authentication request even though it isn't mentioned in the documentation.  To do so, use the `sts_region` parameter.

**Comment by byang-confluent** [comment]
i am hitting the same issue  that `sts_region` doesn't work
```
provider "vault" {
  version = "3.6.0"
  address = "xxx"
  auth_login {
    path   = "xxx"
    method = "aws"
    parameters = {
      role         = "xxx"
      header_value = "xxx"
      sts_region   = "us-west-2"
    }
  }
}
```
```
│ * error making upstream request: received error code 403 from STS: <ErrorResponse xmlns="https://sts.amazonaws.com/doc/2011-06-15/">
│   <Error>
│     <Type>Sender</Type>
│     <Code>SignatureDoesNotMatch</Code>
│     <Message>Credential should be scoped to a valid region. </Message>
│   </Error>
│   <RequestId>233b467a-e4b3-4970-8c9a-3e71d5775eae</RequestId>
│ </ErrorResponse>
```

**Comment by toddmichael** [comment]
> i am hitting the same issue that `sts_region` doesn't work

Also running into this issue with `v3.7.0` of the provider.

Tried adding both `sts_region` **and** `sts_endpoint` (saw a doc that mentioned they need to be used together), but same result.

```
│   <Error>
│     <Type>Sender</Type>
│     <Code>SignatureDoesNotMatch</Code>
│     <Message>Credential should be scoped to a valid region. </Message>
│   </Error>
│   <RequestId>16bfc0c1-3fa6-4db3-a368-d9ac546d3199</RequestId>
│ </ErrorResponse>
```

**Comment by eugenetolan** [comment]
Same issue seen for myself and have tried both sts_region and sts_endpoint

```
Error: Error making API request.
Code: 400. Errors:


* error making upstream request: received error code 403 from STS: <ErrorResponse xmlns="https://sts.amazonaws.com/doc/2011-06-15/">

  <Error>

    <Type>Sender</Type>

    <Code>SignatureDoesNotMatch</Code>

    <Message>Credential should be scoped to a valid region. </Message>

  </Error>

  <RequestId>440f5b21-17ae-48d6-998d-2d06c1e70127</RequestId>

</ErrorResponse>
```

**Comment by byang-confluent** [comment]
my college figure it out that if vault is set to global endpoint, you will need to set sts_region to `us-east-1` to make it work. The region must be set to us-east-1  which the region for global endpoint

**Comment by benashz** [comment]
Fixed in #1618
